src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,228 @@
+Authors: Wrenna Robson
+-/
+
+import analysis.normed_space.basic
+
+/-!
+# Hamming spaces
+
+The Hamming metric counts the number of places two members of a (finite) Pi type
+differ. The Hamming norm is the same as the Hamming metric over additive groups, and
+counts the number of places a member of a (finite) Pi type differs from zero.
+
+This is a useful notion in various applications, but in particular it is relevant
+in coding theory, in which it is fundamental for defining the minimum distance of a
+code.
+
+In this file we define `hamm β`, the type synonym of a Pi type with the Hamming
+distance `hamm_dist` and weight `hamm_wt` attached, and the various instances that arise
+from the properties of these definitions.
+
+-/
+
+/--
+Type synonym for a Pi type which we equip with the Hamming metric, adding all relevant
+instances as needed.
+-/
+def hamm {ι : Type*} (β : ι → Type*) : Type* := Π i, β i
+
+namespace hamm
+
+section
+instance {ι : Type*} (β : ι → Type*) [Π i, inhabited (β i)] : inhabited (hamm β) :=
+⟨λ i, default⟩
+
+local notation `𝓗[` K`,` n`]` := hamm (λ _ : fin n, K)
+
+variables {α ι : Type*} {β : ι → Type*}
+
+/-- `to_hamm` is the identity function to the `hamm` of a type.  -/
+@[pattern] def to_hamm : (Π i, β i) ≃ hamm β := equiv.refl _
+
+/-- `of_hamm` is the identity function from the `hamm` of a type.  -/
+@[pattern] def of_hamm : hamm β ≃ Π i, β i := equiv.refl _
+
+@[simp] lemma to_hamm_symm_eq                 : (@to_hamm _ β).symm = of_hamm := rfl
+@[simp] lemma of_hamm_symm_eq                 : (@of_hamm _ β).symm = to_hamm := rfl
+@[simp] lemma to_hamm_of_hamm (x : hamm β)    : to_hamm (of_hamm x) = x := rfl
+@[simp] lemma of_hamm_to_hamm (x : Π i, β i)  : of_hamm (to_hamm x) = x := rfl
+@[simp] lemma to_hamm_inj {x y : Π i, β i}    : to_hamm x = to_hamm y ↔ x = y := iff.rfl
+@[simp] lemma of_hamm_inj {x y : hamm β}      : of_hamm x = of_hamm y ↔ x = y := iff.rfl
+
+instance [Π i, has_zero (β i)] : has_zero (hamm β) := pi.has_zero
+instance [Π i, has_sub (β i)] : has_sub (hamm β) := pi.has_sub
+instance [Π i, has_scalar α (β i)] : has_scalar α (hamm β) := pi.has_scalar
+instance [has_zero α] [Π i, has_zero (β i)] [Π i, smul_with_zero α (β i)] :
+smul_with_zero α (hamm β) := pi.smul_with_zero _
+instance [Π i, add_monoid (β i)] : add_monoid (hamm β) := pi.add_monoid
+instance [Π i, add_comm_monoid (β i)] : add_comm_monoid (hamm β) := pi.add_comm_monoid
+instance [Π i, add_comm_group (β i)] : add_comm_group (hamm β) := pi.add_comm_group
+instance (α) [semiring α] (β: ι → Type*) [Π i, add_comm_monoid (β i)]
+[Π i, module α (β i)] : module α (hamm β) := pi.module _ _ _
+
+end
+
+/-
+We define `hamm_dist` and `hamm_wt` over Pi types, and will later attach them to our type
+synonym.
+-/
+
+section
+open fintype
+
+variables {α ι : Type*} {β : ι → Type*} [fintype ι] [Π i, decidable_eq (β i)]
+
+/-- The Hamming distance function to the naturals. -/
+
+def hamm_dist (x y : Π i, β i) := card {i // x i ≠ y i}
```

Comment:
```suggestion
/-- The Hamming distance function to the naturals. -/
def hamm_dist (x y : Π i, β i) := card {i // x i ≠ y i}
```
`/--` comments are "attatched" to definitions, so should appear as such in the source code too